astropy/io/fits/hdu/hdulist.py

```diff
@@ -694,14 +694,13 @@ def writeto(self, fileobj, output_verify='exception', overwrite=False,
         # but only if the file doesn't exist.
         fileobj = _File(fileobj, mode='ostream', overwrite=overwrite)
         hdulist = self.fromfile(fileobj)
+        dirname = os.path.dirname(hdulist._file.name)
 
-        for hdu in self:
-            hdu._prewriteto(checksum=checksum)
-            try:
+        with _free_space_check(self, dirname=dirname):
+            for hdu in self:
+                hdu._prewriteto(checksum=checksum)
                 hdu._writeto(hdulist._file)
-            finally:
```

Comment:
The removal of `try-except-finally` here assume that it only tries to catch "out of space" error. Is that always true? What if write fails for other reason? Then it won't clean up properly without the `finally`. Or perhaps you moved the `finally` to L731 below?